keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+#### Stage 2 / Beta
+
+Logging context propagation
+
+1. New `k8s.io/component-base/kontext` package should be implemented which would provide functionalities related to Context handling in Kubernetes:
+    * Attach given logr.Logger instance to existing Context instance.
+    * Get logr.Logger attached to given Context.
+    * Attach new Kubernetes object to Context and attached logr.Logger.
+    * Get Kubernetes object attached to given Context.
+1. All entry points to the application like main() function and HTTP request handlers should be identified. In each of those entry points a new Context should be created if it does not exist already and a module logger should be associated with this context.
+1. If the entry point to the application handles some Kubernetes object like pod or node this object should also be associated with the entry point Context.
+1. Starting from the identified entry points Context containing a logger should be propagated to all places where we log anything. The propagation should be done directly by passing Context as an argument to called functions/methods or indirectly by associating a Context with existing objects passed between functions/methods. In all places where context has been added references to module logger should be replaced with `k8s.io/component-base/kontext.getLogger(context)` calls.
+
+Manual logs verification
+
+1. Detailed documentation for the structured logging best practices in Kubernetes and migration guide for moving from unstructured to structured logs should be prepared.
+1. For each automatically generates logr call following checks should be performed manually:
+    * If log does not provide true value from the developer or user point of view it should be removed.
+    * Log message should be checked if it is meaningful.
+    * Log level should be verified if it correct (Error() vs Info()).
+    * Log verbosity should be verified if it correct..
+    * Log entry fields should be added or changed.
+    * Where it makes sense Kubernetes objects should be attached to context.
+
+#### Stage 3 / GA
+
+1. All places were we log anything in the Kubernetes should be reviewed before reaching this stage and all log messages and fields should be stage enough so that they should not change significantly in the next releases.
+1. `json` log format should become a default.
+
+#### Stage 4 / Deprecation
+
+Cleanup
+
+1. Support for `legacy` logging format should be removed.
+1. klog wrapper package and all of its references should be removed from Kubernetes source code.
+1. All dependencies which are included in Kubernetes and which are currently using klog for logging should be migrated to logr. After this is done klog to logr bridge using klogr should be removed from Kubernetes source code.
+
+
+### User Stories
+
+#### Story 1 - Kubernetes developer working with local logs
+
+TODO(serathius):No major impact when switching from legacy to json format
```

Comment:
Specifically, I think some examples of `sed/awk` --> `jq` might be useful here.  i.e. "this exercise is still just as easy if not easier"